### PR TITLE
MGMT-8859: Refactor agentCluster spec to hold just the ControlPlaneEnpoint

### PR DESCRIPTION
**Pull request by tsorya**
[MGMT-8859](https://issues.redhat.com/browse/MGMT-8859): Refactor agentCluster spec to hold just the ControlPlaneEndpoint

### CODE REVIEW

File: controllers/agentcluster_controller.go [hunk 1]
```diff
@@ -46,6 +50,7 @@ type AgentClusterReconciler struct {
 //+kubebuilder:rbac:groups=capi-provider.agent-install.openshift.io,resources=agentclusters/finalizers,verbs=update
 //+kubebuilder:rbac:groups=hive.openshift.io,resources=clusterdeployments,verbs=get;list;watch;create;update;patch;delete
 //+kubebuilder:rbac:groups=extensions.hive.openshift.io,resources=agentclusterinstalls,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=hypershift.openshift.io,resources=hostedcontrolplanes,verbs=get;list;watch;
```
Comment (avishayt) [inline controllers/agentcluster_controller.go:53]: Will this fail if hypershift isn't installed?

Comment (tsorya) [reply controllers/agentcluster_controller.go:53]: i don't know to say the truth but without it you can't reach the resource

Comment (tsorya) [reply controllers/agentcluster_controller.go:53]: added junk resource and it looks that it passes

File: Makefile [hunk 2]
```diff
@@ -109,7 +111,7 @@ build: generate fmt vet lint ## Build manager binary.
 run: manifests generate fmt vet ## Run a controller from your host.
 	go run ./main.go
 
-docker-build: test ## Build docker image with the manager.
```
Comment (tsorya) [inline Makefile:112]: fix

File: controllers/agentcluster_controller.go [hunk 3]
```diff
@@ -155,6 +221,52 @@ func (r *AgentClusterReconciler) createClusterDeployment(ctx context.Context, lo
...
+	if controlPlane == nil {
+		result := ctrl.Result{}
+		if err != nil {
+			result.Requeue = true
+		}
+		return result, err
+	}
+
+	// Ensure we add a watcher to the control plane object.
+	err = r.addControlPlaneWatcher(ctx, log, controlPlane)
```
Comment (eranco74) [inline controllers/agentcluster_controller.go:259]: Why do we need a watcher here?
We call this method once!

Comment (eranco74) [reply controllers/agentcluster_controller.go:259]: Also, do we think the parameters we get from the control plane might change?

**Review by eranco74 — COMMENTED** [review]
Overall, looks good

File: controllers/agentcluster_controller.go (hunk 3 above)
Comment (tsorya) [reply controllers/agentcluster_controller.go:259]: @avishayt what do you think?

File: controllers/agentcluster_controller.go [hunk 4]
```diff
@@ -132,18 +141,75 @@ func (r *AgentClusterReconciler) updateAgentClusterInstall(ctx context.Context,
 	return ctrl.Result{}, nil
 }
 
-func (r *AgentClusterReconciler) createClusterDeployment(ctx context.Context, log logrus.FieldLogger, agentCluster *capiproviderv1alpha1.AgentCluster) (ctrl.Result, error) {
-	log.Info("Creating clusterDeployment")
+func (r *AgentClusterReconciler) getUnstructuredControlPlane(ctx context.Context, log logrus.FieldLogger,
+	agentCluster *capiproviderv1alpha1.AgentCluster) (*unstructured.Unstructured, error) {
```
Comment (eranco74) [inline controllers/agentcluster_controller.go:145]: Consider renaming this method to getControlPlane and return a struct with the field we require.
You can use the unstractured instance in this function but return a contgrolPlane instance with known fields

Comment (tsorya) [reply controllers/agentcluster_controller.go:145]: you mean that i need to create new struct?

Comment (tsorya) [reply controllers/agentcluster_controller.go:145]: i need object itself too, for the watcher. If we want to drop it than no problem, i can add unstructured to the struct too. 
As you wish but currently i need unstructured  object itself

File: api/v1alpha1/agentcluster_types.go [hunk 5]
```diff
@@ -39,13 +38,6 @@ type AgentClusterSpec struct {
 	// +required
 	ClusterName string `json:"clusterName"`
```
[the remark this answers is not in the document]
Comment (tsorya) [reply api/v1alpha1/agentcluster_types.go:39]: cause it is not part of Control plane

**Comment by eranco74** [comment]
/hold
Unsure how the test report success, it didn't really run
https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_cluster-api-provider-agent/30/pull-ci-openshift-cluster-api-provider-agent-master-e2e-capi-provider-hypershift/1489670496948588544/artifacts/e2e-capi-provider-hypershift/baremetalds-assisted-setup/build-log.txt

**Pull request merged**